types/mempool/sender_nonce.go

```diff
@@ -0,0 +1,199 @@
+package mempool
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+	huandu "github.com/huandu/skiplist"
+)
+
+var (
+	_ Mempool  = (*senderNonceMempool)(nil)
+	_ Iterator = (*nonceMempoolIterator)(nil)
+)
+
+type senderTxs struct {
+	txQueue *huandu.SkipList
+	head    *huandu.Element
+}
+
+func newSenderTxs() senderTxs {
+	return senderTxs{
+		head:    nil,
+		txQueue: huandu.New(huandu.LessThanFunc(txKeyLessNonce)),
+	}
+}
+
+func (s *senderTxs) insert(key txKey, tx sdk.Tx) {
+	s.txQueue.Set(key, tx)
+	s.head = s.txQueue.Front()
```

Comment:
I think we should just track `head` during iteration only.  With this implementation an insert with an open iterator from `Select` could cause re-enumeration of already visited txs.

Reply:
Yes I was just asking this in slack ok I will change it now